### ISSUE TITLE
dory failed to fetch metadata when one of the brokers is down

### ISSUE THREAD

**Issue by shikloshi** [dspeterson/dory#10]
Hey, 

I'm using dory to produce messages to `kafka 0.8.2.1`,

I have two brokers listed in `dory_conf.xml`, something like this:

```xml
    <initialBrokers>
            <broker host="some-broker-host1" port="9092" />
            <broker host="some-broker-host2" port="9092" />
     </initialBrokers>
```

broker1 (`some-broker-host1`) had died and dory started listing these in the log: 

```
2017-07-30T11:29:32.371899000Z dory[6]: Metadata request failed for all known brokers, waiting 6762 milliseconds before retry (1)
2017-07-30T11:29:39.137721000Z dory[6]: Router thread getting metadata from broker some-broker-host2 port 9092
2017-07-30T11:29:39.144926000Z dory[6]: Failed to build metadata structure from response: [src/dory/metadata.cc, 106]; Metadata contains partition with unknown broker ID
2017-07-30T11:29:39.145557000Z dory[6]: Router thread did not get valid metadata response from broker
2017-07-30T11:29:39.145998000Z dory[6]: Router thread getting metadata from broker some-broker-host1 port 9092
2017-07-30T11:29:39.147938000Z dory[6]: Router thread failed to connect to broker for metadata
```

I get the error for `some-broker-host1` (it died and therefor unreachable) but why can't dory fetch the metadata from `some-broker-host2`?

This is the interesting error:
```
2017-07-30T11:29:39.144926000Z dory[6]: Failed to build metadata structure from response: [src/dory/metadata.cc, 106]; Metadata contains partition with unknown broker ID
```

When debugging this, I tried to kill `some-broker-host2` while host1 is still alive, dory behaved the same (host2 was unreachable and for host1 we got the above error), meaning it needed *all* brokers to be available in order to fetch metadata properly. 

Thanks,
Aviad.

**Comment by dspeterson** [comment]
Hi Aviad,

From the output, it looks likely that dory is being too strict in its validation of the metadata response it gets back from the broker.  However I need to confirm this.  I am out of town without access to my development machine until August 7, so my ability to debug the problem is limited until then.  In the meantime, I may still be able to identify the problem if you can help me a bit.  Can you rebuild dory with some debug instrumentation added to the code, reproduce the problem, and tell me what output the instrumentation produces?  Specifically, in src/dory/kafka_proto/metadata/v0/metadata_proto.cc, lines 96-101 contain the following chunk of code:

```c++
    while (reader.NextPartitionInTopic()) {
      int16_t error_code = reader.GetCurrentPartitionErrorCode();
      builder.AddPartitionToTopic(reader.GetCurrentPartitionId(),
          reader.GetCurrentPartitionLeaderId(),
          CanSendToPartition(error_code), error_code);
    }
```

Can you replace that with the following?

```c++
    while (reader.NextPartitionInTopic()) {
      int16_t error_code = reader.GetCurrentPartitionErrorCode();

      try {
        builder.AddPartitionToTopic(reader.GetCurrentPartitionId(),
            reader.GetCurrentPartitionLeaderId(),
            CanSendToPartition(error_code), error_code);
      } catch (const TMetadata::TPartitionHasUnknownBroker &) {
        syslog(LOG_ERR, "=== topic [%s] partition %lu leader %lu error_code %d ===",
            name.c_str(), static_cast<unsigned long>(reader.GetCurrentPartitionId()),
            static_cast<unsigned long>(reader.GetCurrentPartitionLeaderId()),
            static_cast<int>(error_code));
        throw;
      }
    }
```

Likewise, lines 78-83 of src/dory/kafka_proto/metadata/v0/metadata_proto.cc
contain the following chunk of code:

```c++
  while (reader.NextBroker()) {
    name.assign(reader.GetCurrentBrokerHostBegin(),
        reader.GetCurrentBrokerHostEnd());
    builder.AddBroker(reader.GetCurrentBrokerNodeId(), std::move(name),
        reader.GetCurrentBrokerPort());
  }
```

Can you replace that with the following?

```c++
  while (reader.NextBroker()) {
    name.assign(reader.GetCurrentBrokerHostBegin(),
        reader.GetCurrentBrokerHostEnd());
    syslog(LOG_ERR, "=== broker [%s] port %u node id %d ===", name.c_str(),
        static_cast<unsigned>(reader.GetCurrentBrokerPort()),
        static_cast<int>(reader.GetCurrentBrokerNodeId()));
    builder.AddBroker(reader.GetCurrentBrokerNodeId(), std::move(name),
        reader.GetCurrentBrokerPort());
  }
```

This will indicate which topic and partition is causing the problem, which broker the lead replica for that partition lives on, and the error code associated with the partition.  Once you have identified which partition is causing the problem, can you tell me how many replicas that partition has?  I kind of suspect that the partition causing the problem has only a single replica which lives on the crashed broker, but we will see...

Thanks,
Dave

**Comment by shikloshi** [comment]
Thanks for the quick response. 
This is what I came up with:

```
2017-07-31T05:29:41.869024070Z dory[6]: === broker [some-broker-host1] port 9092 node id 1 ===
2017-07-31T05:29:41.869085996Z dory[6]: === topic [kafka-topic] partition 83 leader 18446744073709551615 error_code 5 ===
2017-07-31T05:29:41.869205852Z dory[6]: Failed to build metadata structure from response: [src/dory/metadata.cc, 106]; Metadata contains partition with unknown broker ID
2017-07-31T05:29:41.869216639Z dory[6]: Router thread did not get valid metadata response from broker
```

But when running `kafka-topics.sh --describe..`, I see that this partition has two replicas:
```
	Topic: kafka-topic	Partition: 83	Leader: -1	Replicas: 2	Isr:
```

**Comment by shikloshi** [comment]
Looks like you were right, some of the topics was not replicated (replication factor of 1) and after broker died had troubles finding new leader for the partition. 

Thanks for your help.

**Comment by dspeterson** [comment]
Hi Aviad,

I'm glad you were able to determine what is going on.  By the way, the debug instrumentation shows 18446744073709551615 for the leader ID because it's actually a signed value and I casted it to an unsigned by mistake, and that's what -1 looks when displayed as a 64-bit unsigned integer.  Although you identified the problem with the unreplicated partitions, you did also discover a problem with dory.  In a case where dory gets back a metadata response like what you are seeing, it should be more tolerant, and simply log a warning message and mark the partition as unavailable rather than rejecting the entire response.  When I get back from my trip, I will implement a fix.  Also I will look at dory's metadata processing code in more depth and see if there other cases where it should be more tolerant about the metadata responses it accepts from the brokers.  Thanks very much for identifying this issue. :-)

Dave

**Comment by shikloshi** [comment]
Hey Dave,

Yes, exactly this is what I though I saw.
Thanks for your help debugging this, enjoy the rest of the trip and I will follow up here for the fix!

Thanks,
Aviad.

**Comment by dspeterson** [comment]
fixed in version 2.0.5
